File: Interview Challenge/C++/missing-number.cpp
// https://codefights.com/interview/t8iNoBtDf5XSiwHiF
int missingNumber(std::vector<int> a) {
    int n = a.size();
    int r = 0;
    if(n == 0) return r;
    if(n == 1 && a[0] == 0) return 1;

    sort(a.begin(), a.end());
    for(int i = 1; i < n; i++) {
        if(a[i-1] + 1 != a[i]) {
            r = a[i] - 1;
            if(r == a[i-1]) r = a[0] - 1;
            if(r < 0) r = a[n-1] + 1;
            return r;
        }
        else if(a[i-1] + 1 == a[i] && i == n-1) {
            if(a[0] - 1 < 0) return a[n-1] + 1;
            else return a[0] - 1;
        }
    }

    return r;
}

File: Challenge/Javascript/missingNumber.js
// https://app.codesignal.com/challenge/p4qArskGsNtf6ecdP

missingNumber = a =>
{
    // Another method
    // s = 0
    // a.map(e => s += e + 1)
    // l = a.length + 1
    // return (r = (l * (l+1)) / 2 % s - 1) < 0 ? l - 1 : r

    i = 0
    for(e of a.sort((a,b) => a - b))
        if(e != i++) {
            --i
            break
        }
    return i
}
